### ISSUE TITLE
Runtime Error when trying to run native=True

### ISSUE THREAD

**Issue by swrpug** [zauberzeug/nicegui#1841]
### Description

I'm getting a Fork vs Spawn error when attempting to run a native app on Ubuntu.

**Sample Application (from documentation)**
```py
from nicegui import app, ui

app.native.window_args['resizable'] = False
app.native.start_args['debug'] = True

ui.label('app running in native mode')
ui.button('enlarge', on_click=lambda: app.native.main_window.resize(1000, 700))

ui.run(native=True, window_size=(400, 300), fullscreen=False)

```

Invoked from command line from within a virtual environment:  python main.py

**Trackback**
```
Traceback (most recent call last):
  File "/mnt/ext3tbdrive/sandbox/try-nicegui/main.py", line 15, in <module>
    main()
  File "/mnt/ext3tbdrive/sandbox/try-nicegui/main.py", line 11, in main
    ui.run(native=True, window_size=(400, 300), fullscreen=False)
  File "/mnt/ext3tbdrive/sandbox/try-nicegui/env/lib/python3.11/site-packages/nicegui/run.py", line 180, in run
    ChangeReload(config, target=globals.server.run, sockets=[sock]).run()
  File "/mnt/ext3tbdrive/sandbox/try-nicegui/env/lib/python3.11/site-packages/uvicorn/supervisors/basereload.py", line 49, in run
    self.startup()
  File "/mnt/ext3tbdrive/sandbox/try-nicegui/env/lib/python3.11/site-packages/uvicorn/supervisors/basereload.py", line 85, in startup
    self.process.start()
  File "/home/tim/.pyenv/versions/3.11.6/lib/python3.11/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "/home/tim/.pyenv/versions/3.11.6/lib/python3.11/multiprocessing/context.py", line 288, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "/home/tim/.pyenv/versions/3.11.6/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/home/tim/.pyenv/versions/3.11.6/lib/python3.11/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/home/tim/.pyenv/versions/3.11.6/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/home/tim/.pyenv/versions/3.11.6/lib/python3.11/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
  File "/home/tim/.pyenv/versions/3.11.6/lib/python3.11/multiprocessing/synchronize.py", line 107, in __getstate__
    raise RuntimeError('A SemLock created in a fork context is being '
RuntimeError: A SemLock created in a fork context is being shared with a process in a spawn context. This is not supported. Please use the same context to create multiprocessing objects and Process.
```

OS: Ubuntu 20.04
Python: 3.11.6
Pyenv: 2.3.29-5-g62e2e9c7
nicegui: 1.3.17
pywebview: 4.3.3

**Comment by rodja** [comment]
Strange. On Mac it's working fine. @JensOgorek could you check if the error is reproducible on your system?

**Comment by JensOgorek** [comment]
@rodja I just saw this. Can confirm. It starts normally, but clicking on enlarge causes connection lost at nicegui version 1.3.17 and on 1.4.1.

**Comment by muratbedir** [comment]
On the Windows 11, it requires webview to run 

![image](https://github.com/zauberzeug/nicegui/assets/48950062/65873bd0-fcf9-491b-9115-02bb3c6c9b61)


microsoft code log


PowerShell 7.3.8
PS C:\Users\mbedir\Desktop\niceekran>  & 'C:\Program Files\Python311\python.exe' 'c:\Users\mbedir\.vscode\extensions\ms-python.python-2023.18.0\pythonFiles\lib\python\debugpy\adapter/../..\debugpy\launcher' '61418' '--' 'C:\Users\mbedir\Desktop\niceekran\destopapp.py'
PS C:\Users\mbedir\Desktop\niceekran>  c:; cd 'c:\Users\mbedir\Desktop\niceekran'; & 'C:\Program Files\Python311\python.exe' 'c:\Users\mbedir\.vscode\extensions\ms-python.python-2023.18.0\pythonFiles\lib\python\debugpy\adapter/../..\debugpy\launcher' '61442' '--' 'C:\Users\mbedir\Desktop\niceekran\destopapp.py'

Native mode is not supported in this configuration.
Please run "pip install pywebview" to use it.
PS C:\Users\mbedir\Desktop\niceekran> pip install pywebview  
Defaulting to user installation because normal site-packages is not writeable
DEPRECATION: Loading egg at c:\program files\python311\lib\site-packages\vboxapi-1.0-py3.11.egg is deprecated. pip 23.3 will enforce this behaviour change. A possible replacement is to use pip for package installation..
Collecting pywebview
  Obtaining dependency information for pywebview from https://files.pythonhosted.org/packages/94/eb/fec1105c0e3b459266637bdc867bd064619eddb170dc0c9d12f20ad7690c/pywebview-4.4.1-py3-none-any.whl.metadata
  Downloading pywebview-4.4.1-py3-none-any.whl.metadata (4.0 kB)
Collecting proxy-tools (from pywebview)
  Downloading proxy_tools-0.1.0.tar.gz (3.0 kB)
  Preparing metadata (setup.py) ... done
Requirement already satisfied: bottle in c:\users\mbedir\appdata\roaming\python\python311\site-packages (from pywebview) (0.12.23)
Requirement already satisfied: typing-extensions in c:\users\mbedir\appdata\roaming\python\python311\site-packages (from pywebview) (4.8.0)
Collecting pythonnet (from pywebview)
  Obtaining dependency information for pythonnet from https://files.pythonhosted.org/packages/12/28/2ecca48b313c436eb0b194b74b5a70dd88f2700d79bce5424799ffc82ede/pythonnet-3.0.3-py3-none-any.whl.metadata
  Downloading pythonnet-3.0.3-py3-none-any.whl.metadata (6.6 kB)
Collecting clr-loader<0.3.0,>=0.2.6 (from pythonnet->pywebview)
  Obtaining dependency information for clr-loader<0.3.0,>=0.2.6 from https://files.pythonhosted.org/packages/b3/1a/2801c4e3888aac8fcd4d88d32a1155a778d1c01ca040ae8137d433a39cc6/clr_loader-0.2.6-py3-none-any.whl.metadata
  Downloading clr_loader-0.2.6-py3-none-any.whl.metadata (1.4 kB)
Requirement already satisfied: cffi>=1.13 in c:\users\mbedir\appdata\roaming\python\python311\site-packages (from clr-loader<0.3.0,>=0.2.6->pythonnet->pywebview) (1.15.1)
Requirement already satisfied: pycparser in c:\users\mbedir\appdata\roaming\python\python311\site-packages (from cffi>=1.13->clr-loader<0.3.0,>=0.2.6->pythonnet->pywebview) (2.21)
Downloading pywebview-4.4.1-py3-none-any.whl (405 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 405.9/405.9 kB 1.9 MB/s eta 0:00:00
Downloading pythonnet-3.0.3-py3-none-any.whl (290 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 291.0/291.0 kB 3.0 MB/s eta 0:00:00
Downloading clr_loader-0.2.6-py3-none-any.whl (51 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.3/51.3 kB 1.3 MB/s eta 0:00:00
Building wheels for collected packages: proxy-tools
  Building wheel for proxy-tools (setup.py) ... done
  Created wheel for proxy-tools: filename=proxy_tools-0.1.0-py3-none-any.whl size=2913 sha256=8b3a53f5a99ad97b8c2394db2a73ef5523255a24e849ee13b0db7091d9012d90
  Stored in directory: c:\users\mbedir\appdata\local\pip\cache\wheels\2b\5f\96\7e96e7cbf06e426e168a6891a6ca2226787cc79fc816830dd4
Successfully built proxy-tools
Installing collected packages: proxy-tools, clr-loader, pythonnet, pywebview
Successfully installed clr-loader-0.2.6 proxy-tools-0.1.0 pythonnet-3.0.3 pywebview-4.4.1

[notice] A new release of pip is available: 23.2.1 -> 23.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip
PS C:\Users\mbedir\Desktop\niceekran>  c:; cd 'c:\Users\mbedir\Desktop\niceekran'; & 'C:\Program Files\Python311\python.exe' 'c:\Users\mbedir\.vscode\extensions\ms-python.python-2023.18.0\pythonFiles\lib\python\debugpy\adapter/../..\debugpy\launcher' '61479' '--' 'C:\Users\mbedir\Desktop\niceekran\destopapp.py'

0.01s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
0.01s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
NiceGUI ready to go on http://localhost:8000 (install netifaces to show all IPs and speedup this message)
[pywebview] Using WinForms / Chromium
PS C:\Users\mbedir\Desktop\niceekran>  c:; cd 'c:\Users\mbedir\Desktop\niceekran'; & 'C:\Program Files\Python311\python.exe' 'c:\Users\mbedir\.vscode\extensions\ms-python.python-2023.18.0\pythonFiles\liions\ms-python.python-2023.18.0\pythonFiles\lib\python\debugpy\adapter/../..\debugpy\launcher' '6154ions\ms-python.python-2023.18.0\pythonFiles\lib\python\debugpy\adapter/../..\debugpy\launcher' '61543' '--' 'C:\Users\mbedir\Desktop\niceekran\destopapp.py'
0.01s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
0.01s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
NiceGUI ready to go on http://localhost:8000 (install netifaces to show all IPs and speedup this message)
[pywebview] Using WinForms / Chromium

**Comment by rodja** [comment]
@muratbedir Yes the message 

```
Native mode is not supported in this configuration.
Please run "pip install pywebview" to use it.
```

Is expected because we made pywebview an optional dependency. It has nothing to do with this issue.

**Comment by fdawnxl** [comment]
I also encountered the same situation.Environment related information is as follows:

- OS Version:ArchLinux Linux 6.5.9-arch2-1
- Python Version: 3.11.5
- NiceGUI Version:1.4.1

When I execute the following code in the venv virtual environment, python will report an error:

```
from nicegui import app, ui

app.native.window_args['resizable'] = False
app.native.start_args['debug'] = True

ui.label('app running in native mode')
ui.button('enlarge', on_click=lambda: app.native.main_window.resize(1000, 700))

ui.run(native=True, window_size=(400, 300), fullscreen=False)


Traceback(part):
  File "/usr/lib/python3.11/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
  File "/usr/lib/python3.11/multiprocessing/synchronize.py", line 107, in __getstate__
    raise RuntimeError('A SemLock created in a fork context is being '
RuntimeError: A SemLock created in a fork context is being shared with a process in a spawn context. This is not supported. Please use the same context to create multiprocessing objects and Process.

```
Because pip cannot be used to directly install python packages under ArchLinux (it will conflict with pacman), and there is no packaged python-nicegui package, so I have been developing in a virtual environment. I used to think that the error was due to executing NiceGUI in a virtual environment, but after reading [this article](https://stackoverflow.com/questions/69332196/mutiprocessing-with-spawn-context-cannot-access-shared-variables-in-linux), I tend to think that it is due to incorrect configuration of multiprocessing.(However, I still recommend that you directly use the system python environment to execute the NiceGUI code in a Linux distribution that can use pip to determine whether this problem can be reproduced.)

**Comment by dev-robasi** [comment]
It works in windows via a virtual env. Had same error with wsl-ubuntu in a virtual env.

**Comment by nimmadev** [comment]
same issue on 
system: Linux  6.5.13-7-MANJARO 
python: 3.12.1
python installed with pyenv
using virtual env

**Comment by ahdeddd** [comment]
Same issue here on Manjaro Linux. I installed pywebview with --break-system-packages

**Comment by yuanxion** [comment]
1."RuntimeError: A SemLock created in a fork context is being shared with a process in a spawn context." issue is not happened when running the following sample on my host with Python 3.10 and nicegui (both 1.3.17 & 1.4.12).

// On host
Operating System: Ubuntu 22.04.3 LTS              
Kernel: Linux 6.5.0-14-generic
Python: 3.10.13
nicegui: 1.3.17/1.4.12

```
from nicegui import app, ui

app.native.window_args['resizable'] = False
app.native.start_args['debug'] = True

ui.label('app running in native mode')
ui.button('enlarge', on_click=lambda: app.native.main_window.resize(1000, 700))

ui.run(native=True, window_size=(400, 300), fullscreen=False)
```

2.But I will have another issue when running nicegui with "native=True": **Connection terminated unexpectedly**

![image](https://github.com/zauberzeug/nicegui/assets/96522341/09021ce7-6f7e-4ed8-8e93-6dd7ec739eae)

**Comment by electronstudio** [comment]
Add to the top of your code:
```
import multiprocessing
multiprocessing.set_start_method("spawn", force=True)
```

On Linux I also had to `pip install qtpy pyqt5 PyQtWebEngine`

**Comment by fdawnxl** [comment]
I've tried the method above, but NicuGUI still throws the same error as before. I think this might be a venv issue. So I created an ubuntu virtual machine (with Gnome desktop) and installed NiceGUI and pywebview in the system python environment. After I installed the relevant dependencies of pywebview, I directly executed the sample code on the homepage. Surprisingly, the program was executed successfully:
![Uploading Screenshot from 2024-02-15 00-01-24.png…]()
Since I did not use the same system and kernel as the host machine(Arch Linux, linux-zen) when I created the virtual machine(Ubuntu, linux), I cannot find the problem by controlling variables for the time being. However, this may indicate that there is no problem with the code of NiceGUI itself.
By the way, when using pywebview, you may get an error due to missing dependencies. You can refer to [this URL](https://pywebview.flowrl.com/2.4/guide/installation.html#dependencies) to solve this problem.

**Comment by initiateit** [comment]
Using:

```from nicegui import app, ui
import multiprocessing
multiprocessing.set_start_method("spawn", force=True)

app.native.window_args['resizable'] = False
app.native.start_args['debug'] = True

ui.label('app running in native mode')
ui.button('enlarge', on_click=lambda: app.native.main_window.resize(1000, 700))

ui.run(native=True, window_size=(400, 300), fullscreen=False)
```
Getting Error:

`RuntimeError: A SemLock created in a fork context is being shared with a process in a spawn context. This is not supported. Please use the same context to create multiprocessing objects and Process.`

Operating system is: **Ubuntu 23.10**

Installed via:
```python3 -m pip install pywebview nicegui qtpy pyqt5 PyQtWebEngine```

**Comment by gsal** [comment]
Getting same issue: 

Rocky 9.3
Python 3.11.9
Nicegui 1.4.25
pywebview 5.1

BUT, I cannot install what I need into the system, so, I am working out of a conda environment.

**Comment by StephanieSunshine** [comment]
Same issue EndeavourOS x86_64

pywebview: 5.1
nicegui: 1.4.26
python: 3.12.3

Tried to install python-pywebview from AUR to possibly bring in any dependencies missing. Still no luck.
[AUR](https://aur.archlinux.org/packages/python-pywebview)

**Comment by knoppmyth** [comment]
Here is how I got it working using venv (Using Arch Linux):

```
python -m venv venv
source venv/bin/activate
pip install nicegui pywebview qtpy pyqt5 PyQtWebEngine
```
Here is the code I was able to run without issue.
```
import multiprocessing
multiprocessing.set_start_method("spawn", force=True)
from nicegui import app, ui


app.native.window_args['resizable'] = False
app.native.start_args['debug'] = True

ui.label('app running in native mode')
ui.button('enlarge', on_click=lambda: app.native.main_window.resize(1000, 700))

ui.run(native=True, window_size=(400, 300), fullscreen=False)
```

**Note:** I only saw the error other reported above if:
```
import multiprocessing
multiprocessing.set_start_method("spawn", force=True)
```
Came after:
```
from nicegui import app, ui
```

**Comment by frankhuurman** [comment]
I've kind of given up on running my nicegui app in native mode on Ubuntu 24.04.1.
Also had the SemLock issue.
I did get it running with the suggested packages and multiprocessing import from @knoppmyth but the app scales extremely weird and small even if I stretch it to the full size of the screen.

**Comment by falkoschindler** [comment]
[pywebview 5.3](https://github.com/r0x0r/pywebview/releases/tag/5.3) just got released. Can anyone check if the problem persists?

**Comment by Qhawaq** [comment]
Yes, problem persists ... but it is a VSCode terminal problem. 
If you run program using  linux standard terminal ( shell ) the problem 
doesn't exists. The only warning appears at the end of program execution: 

**_/usr/lib/python3.12/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 6 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '_**

I use Ubuntu 24.04.1, and pywebview 5.3.2 under venv , and I follow the **knoppmyth** solution .

In VScode environment I installed : **pywebview qtpy pyqt5 PyQtWebEngine pycairo PyGObject** to try to solve problem
but it persists . 

A little quirk could be used to solve the first main exception of the program try to unset  the variable GTK_PATH in VSCode terminal, the first exception disappears, but some others problems arises after this quirk.

In VsCode terminal the same program said: 

**NiceGUI ready to go on http://localhost:8000
[pywebview] Using GTK
Gtk-Message: 10:35:45.385: Failed to load module "canberra-gtk-module"
/usr/lib/x86_64-linux-gnu/webkit2gtk-4.1/WebKitNetworkProcess: symbol lookup error: /snap/core20/current/lib/x86_64-linux-gnu/libpthread.so.0: undefined symbol: __libc_pthread_init, version GLIBC_PRIVATE
Gtk-Message: 10:35:45.649: Failed to load module "canberra-gtk-module"
Gtk-Message: 10:35:45.650: Failed to load module "canberra-gtk-module"
Gtk-Message: 10:35:45.673: Failed to load module "canberra-gtk-module"
Gtk-Message: 10:35:45.675: Failed to load module "canberra-gtk-module"
/usr/lib/x86_64-linux-gnu/webkit2gtk-4.1/WebKitNetworkProcess: symbol lookup error: /snap/core20/current/lib/x86_64-linux-gnu/libpthread.so.0: undefined symbol: __libc_pthread_init, version GLIBC_PRIVATE
ERROR: WebKit encountered an internal error. This is a WebKit bug.
./Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp(562) : internallyFailedLoadTimerFired**

So the problem seems to be located only in the VSCode environment and not in venv environment used by python and requested by Ubuntu .

Here you can see the correct execution of **knopmyth** snippet under linux shell terminal

![Schermata del 2024-11-06 10-46-01](https://github.com/user-attachments/assets/b5568067-44ed-4eda-8329-431a541ed58f)

**Comment by sagarbehere** [comment]
I faced the exact same error when running with `ui.run(native=True)` on Ubuntu 24.04 while running inside a python venv in a terminal window. The following worked to fix the error

Step 1: Adding the following lines at the very start of the python script, before `from nicegui import ui`

```python
import multiprocessing
multiprocessing.set_start_method("spawn", force=True)
``` 

Step 2: (I had installed pywebview in the venv with `pip install pywebview`). I had to additionally do `pip install pywebview[qt]`

After that, I was able to run the nicegui app as a native app with no weird appearance issues.

**Comment by frankhuurman** [comment]
@sagarbehere 

installed the same packages as you but on Arch Linux & Python 3.13, even if it has the "qt" libraries installed pywebview still seems to search for the GTK libs:

```python
(venv) [frank@codingthing pyprojects]$ python app.py
NiceGUI ready to go on http://localhost:8000
[pywebview] GTK cannot be loaded
Traceback (most recent call last):
  File "/home/frank/pyprojects/venv/lib/python3.13/site-packages/webview/guilib.py", line 37, in import_gtk
    import webview.platforms.gtk as guilib
  File "/home/frank/pyprojects/venv/lib/python3.13/site-packages/webview/platforms/gtk.py", line 20, in <module>
    import gi
ModuleNotFoundError: No module named 'gi'
(venv) [frank@codingthing pyprojects]$ /usr/lib/python3.13/multiprocessing/resource_tracker.py:276: UserWarning: resource_tracker: There appear to be 6 leaked semaphore objects to clean up at shutdown: {'/mp-vq207iti', '/mp-64bdpzsk', '/mp-0e4qk06e', '/mp-5hddrepa', '/mp-cbmscziu', '/mp-caime8fw'}
  warnings.warn(
```

And have the same leaked semaphore objects warning as you.

But besides those warnings, it does seem to work now with native mode so thanks for that 😄 
Will have to build something more complex to see if that works!

**Comment by evnchn** [comment]
Was review issues and heard "leaked semaphore" multiple times in this thread. 

So, I suggest going to NiceGUI 2.16.0 for https://github.com/zauberzeug/nicegui/pull/4653

**Comment by MrGibus** [comment]
I managed to get this working.

I'm running Fedora i3 with Nvidia Proprietary drivers and Picom compositor in a venv with python3.11

Here's what I did:
## Update to newest version of nicegui
`pip install --upgrade nicegui`
## Install GTK pygobject
I then got an error that there was no module named 'gi' -> See @frankhuurman response above
Googling this -> GTK `GObject Introspection` python bindings are not working so I tried installing these
`pip install pygobject` This did not work for me because of a meson build error.
## Install dev tools if above doesn't work
Double check dependencies are installed - `dnf install python3-devel cairo-devel gobject-introspection-devel`  
If you're on ubuntu or something else you'll need to look up the equivalent libraries
*Make sure to install the correct version of python3-devel (or ubuntu equivalent)  for you venv* - In my case I was a bit of an idiot because my venv is using 3.11 and system python is 3.13. So for me I had to install `sudo dnf install python3.11-devel`
Meson will fail to find python otherwise
After this re-install `pip install --force-reinstall --no-cache-dir pycairo pygobject`
## Additional issues
I also had a GBM buffer issue -> This is GTK nvidia thing.  To get it working I had to run 
`WEBKIT_DISABLE_COMPOSITING_MODE=1 GDK_BACKEND=x11 python app.py`

**Comment by falkoschindler** [comment]
pywebview 6.1 just got released.
Can anyone check if the problem persists after upgrading pywebview as well as NiceGUI? Thanks!

**Comment by knoppmyth** [comment]
Yes, the issue still exist. Still works as noted above https://github.com/zauberzeug/nicegui/issues/1841#issuecomment-2199268613

**Comment by frankhuurman** [comment]
Yup, still exists.
Python 3.13.7 with a fresh virtual env and NiceGUI version 3.2.0 first points me to install pywebview to be able to use native mode.
After installing that and running again it tells me
`webview.errors.WebViewException: You must have either QT or GTK with Python extensions installed in order to use pywebview.`

Then after installing PyGObject and running again I get the dreaded semlock error:
```
RuntimeError: A SemLock created in a fork context is being shared with a process in a spawn context. This is not supported. Please use the same context to create multiprocessing objects and Process.
```

NiceGUI native mode on Linux only seems to work if you put this somewhere in your imports, for me it does not matter if it's before or after importing nicegui:

```py
import multiprocessing
multiprocessing.set_start_method('spawn', force=True)
```

**Comment by evnchn** [comment]
Unable to recreate `RuntimeError: A SemLock created in a fork context is being shared with a process in a spawn context. This is not supported. Please use the same context to create multiprocessing objects and Process.`

I am using `pip install nicegui pywebview[gtk]` in a freshly created venv, Debian 12. 

@ everyone are you Qt or GTK? I suspect a difference. 

Regardless I do have this error though:

<img width="737" height="486" alt="Image" src="https://github.com/user-attachments/assets/0a7db14f-24a4-4703-a5d6-3ea3dcbcf7fd" />

And also:

<img width="1160" height="167" alt="Image" src="https://github.com/user-attachments/assets/deb411f3-e4a6-4d66-a47f-fe5335c3dd05" />

So yeah, not seamless, but cannot recreate the critical issue here.

**Comment by BanDroid** [comment]
for summary:

1. add this in your main script on the first line before you import nicegui.

```python
import multiprocessing
multiprocessing.set_start_method("spawn", force=True)
```

2. install necessary qt library for pywebview.

```bash
pip install pywebview[qt]
```

**Comment by evnchn** [comment]
So is it that only qt need to set the spawn mode?

I may soon add a warning to the user but I want to ensure that what I am warning is precise and do not over-warn.

**Comment by MrGibus** [comment]
Tested on new install:
-> Message to install pywebview
Installed via `pip install pywebview`
-> SemLock error above:
fixed with `multiprocessing.set_start_method("spawn", force=True)`
ModuleNotFoundError: .. Need QT or GTK with python extensions insalled

Here I found that both:
`pip install pywebview[gtk]` works, and
`pip install pywebview[qt]` per above works.

Testing previous install. It should be noted I switched from Nvidia to AMD recently. For the code running previously I no longer need to run them with `WEBKIT_DISABLE_COMPOSITING_MODE=1 GDK_BACKEND=x11 python app.py` I can just run the program normally. Though it still requires the mp line above

**Comment by falkoschindler** [comment]
Thanks for the thorough fresh-install walkthrough @MrGibus\! This confirms that the `multiprocessing.set_start_method("spawn")` workaround is needed on Linux regardless of the GUI backend (both Qt and GTK).

This suggests we should handle this on the NiceGUI side — e.g. automatically setting the start method to `"spawn"` when `native=True` is used on Linux, or at minimum emitting a clear error message pointing users to the fix instead of letting them hit the cryptic SemLock traceback.

**Comment by evnchn** [comment]
> multiprocessing.set_start_method("spawn") workaround is needed on Linux regardless of the GUI backend (both Qt and GTK).

@falkoschindler well at https://github.com/zauberzeug/nicegui/issues/1841#issuecomment-3855279380 I can run native mode without the workaround. It adds to the reason alongside the fact that it is a breaking change, behind why I would advise against applying this for all users. 

A more meaningful error message is good though.

**Comment by evnchn** [comment]
*Drafted by Claude Code working with @evnchn — empirical reproduction across distros and Python versions on the homelab; presenting findings only, no fix recommendation here.*

**TL;DR:** the SemLock error is **a CPython interpreter version cutoff, not a distro or backend issue**. The check that produces this error was added in **CPython 3.11.5** (2023-08-24, [python/cpython#77377](https://github.com/python/cpython/issues/77377), backport [#108378](https://github.com/python/cpython/pull/108378)) — note this was an *improvement* in CPython, not a regression: the bug is latent in NiceGUI's code path, the new check just made it loud (`RuntimeError`) instead of a silent segfault in the spawn child. On the same machine, with the same NiceGUI / pywebview / OS, switching from Python **3.11.4** to **3.11.5** flips the result from "works" to `SemLock` `RuntimeError`. That single fact reconciles every contradictory report in this thread — distros and backends only correlate with whichever Python version they happen to ship.

<details><summary>Per-Python-version sweep (held everything else constant)</summary>

Methodology: spin up Docker containers, install latest `nicegui` (3.11.1) + `pywebview` (6.2.1) in a fresh venv, run the example from the issue body with `native=True` (default `reload=True`). Wrap with `timeout 12s` so the "works" case (which blocks in the event loop without a display) is distinguishable from the "errors out fast" case.

**Pure CPython sweep on `python:X.Y-slim` images, GTK backend, no spawn fix:**

| image | actual Python | result |
| --- | --- | --- |
| `python:3.10-slim` | 3.10.20 | works |
| `python:3.11-slim` | **3.11.15** | **`SEMLOCK_FIRED`** |
| `python:3.12-slim` | 3.12.13 | **`SEMLOCK_FIRED`** |
| `python:3.13-slim` | 3.13.13 | **`SEMLOCK_FIRED`** |

So the cutoff is somewhere inside the 3.11.x patch series. Pinning the boundary precisely with `uv python install` on the same `ubuntu:22.04` base:

| Python | result |
| --- | --- |
| 3.11.4 | works (timeout in event loop, exit 137) |
| **3.11.5** | **`SEMLOCK_FIRED`** |

That's the boundary. Aligns with [python/cpython#77377](https://github.com/python/cpython/issues/77377) — *"multiprocessing: serialization must ensure that contexts are compatible (the same)"* — fixed by [PR #107275](https://github.com/python/cpython/pull/107275) (commit `1700d34d314f`, by @albanD), backported to 3.11 as [PR #108378](https://github.com/python/cpython/pull/108378) (commit `34ef75d3ef55`), first shipped in **CPython 3.11.5** on 2023-08-24. **Not** backported to 3.10, which matches the 3.10.20 result above. A follow-up false-positive correction landed days later as [gh-108520](https://github.com/python/cpython/issues/108520) (commit `385b1952f081`), also in 3.11.5.

What the upstream fix actually changed: before 3.11.5, passing a fork-context `SemLock` to a spawn-context `Process` silently pickled a handle that was invalid in the spawn child — the child would segfault or corrupt shared state. The fix turns that silent crash into the clear `RuntimeError` we now see at pickle time. So this latent bug existed all along; 3.11.5 just made it visible.

**Distro × backend × spawn-fix matrix** (Python pinned per row, NiceGUI 3.11.1, pywebview 6.2.1, default `ui.run(native=True)` with `reload=True`):

| distro | backend | spawn fix | Python | result |
| --- | --- | --- | --- | --- |
| ubuntu:24.04 | qt | no | 3.12.3 | **`SEMLOCK_FIRED`** |
| ubuntu:24.04 | qt | yes | 3.12.3 | works |
| ubuntu:24.04 | gtk | no | 3.12.3 | **`SEMLOCK_FIRED`** |
| ubuntu:24.04 | gtk | yes | 3.12.3 | works |
| debian:12 | gtk | no | **3.11.2** | works |
| debian:12 | qt | no | **3.11.2** | works |
| ubuntu:22.04 | qt | no | 3.10.12 | works |
| ubuntu:22.04 | gtk | no | 3.10.12 | works |

Both Qt and GTK behave identically on the same Python version — backend is not the variable. Distro maps directly to the Python version the distro ships, nothing more:

| reporter | environment | mapped Python | matches my result |
| --- | --- | --- | --- |
| @yuanxion ([here](https://github.com/zauberzeug/nicegui/issues/1841#issuecomment-1907048554)) | Ubuntu 22.04 | 3.10 | works ✓ |
| @evnchn ([here](https://github.com/zauberzeug/nicegui/issues/1841#issuecomment-3855279380)) | Debian 12 + GTK | 3.11.2 | works ✓ |
| @swrpug (OP) | Ubuntu 20.04 + pyenv | 3.11.6 | fires ✓ |
| @initiateit, @MrGibus, @frankhuurman, @StephanieSunshine, @Aikufurr, @Qhawaq et al. | Arch / EndeavourOS / Manjaro / Fedora / Ubuntu 23.10+ / Ubuntu 24.04 | 3.11.5+ or 3.12+ | fires ✓ |

`reload=False` also sidesteps the SemLock on Python 3.12 (no `ChangeReload`, no fork→spawn pickle), confirmed empirically — that's a third workaround alongside the `set_start_method('spawn')` and the older-Python paths.

**Why backend is a red herring**: the `SemLock` is in the `multiprocessing.Queue` / `Pipe` / `Event` objects NiceGUI creates for its native-mode IPC, regardless of whether pywebview ends up using GTK or Qt to render the actual window. Reporters who said "QT works after `set_start_method`" succeeded because of the spawn fix, not because Qt is special; same for GTK reporters who succeeded.

</details>

That's the empirical picture — no fix proposed in this comment, just laying out the version cutoff so the discussion isn't relitigating distros and backends.

---

**Changelog**

- *Edit 2026-05-09*: replaced "regression" with "cutoff" throughout, and clarified in the TL;DR that CPython 3.11.5 was an *improvement* in CPython (silent segfault in spawn child → loud `RuntimeError` at pickle time), not the interpreter going backwards. The latent fault is on NiceGUI's side; the new check just made it visible.
